internal/gcsx/client_readers/range_reader_test.go

```diff
@@ -355,3 +370,145 @@ func (t *rangeReaderTest) Test_invalidateReaderIfMisalignedOrTooSmall() {
+
+			assert.Error(t.T(), err)
+			assert.Zero(t.T(), n)
+			assert.Nil(t.T(), t.rangeReader.reader)
+			assert.Equal(t.T(), int64(-1), t.rangeReader.start)
+			assert.Equal(t.T(), int64(-1), t.rangeReader.limit)
+			assert.Equal(t.T(), tc.readHandle, t.rangeReader.readHandle)
+		})
+	}
+}
+
+func (t *rangeReaderTest) Test_ReadAt_PropagatesCancellation() {
+	// Set up a blocking reader
+	finishRead := make(chan struct{})
+	blocking := &blockingReader{c: finishRead}
+	rc := io.NopCloser(blocking)
+	// Assign it to the rangeReader
+	t.rangeReader.reader = &fake.FakeReader{ReadCloser: rc}
+	t.rangeReader.start = 0
+	t.rangeReader.limit = 2
+	// Track cancel invocation
+	cancelCalled := make(chan struct{})
+	t.rangeReader.cancel = func() { close(cancelCalled) }
+	// Controlled context
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// Channel to track read completion
+	readReturned := make(chan struct{})
+
+	go func() {
+		buf := make([]byte, 2)
+		_, _ = t.rangeReader.ReadAt(ctx, &gcsx.GCSReaderRequest{
+			Buffer:    buf,
+			Offset:    0,
+			EndOffset: 2,
+		})
+		close(readReturned)
+	}()
+
+	// Wait a bit to ensure ReadAt is blocking
+	select {
+	case <-readReturned:
+		t.T().Fatal("Read returned early — cancellation did not propagate properly.")
+	case <-time.After(10 * time.Millisecond):
+		// OK: Still blocked
+	}
+	// Cancel the context to trigger cancellation
+	cancel()
+	// Expect rr.cancel to be called
+	select {
+	case <-cancelCalled:
+		// Pass
+	case <-time.After(100 * time.Millisecond):
+		t.T().Fatal("Expected rr.cancel to be called on ctx cancellation.")
+	}
+	// Unblock the reader so the read can complete
+	close(finishRead)
+	// Ensure read completes
+	select {
+	case <-readReturned:
+		// Pass
+	case <-time.After(100 * time.Millisecond):
+		t.T().Fatal("Expected read to return after unblocking.")
+	}
+}
+
+// Note that this test triggers a race condition: it might receive readDone first,
+// or it might receive ctx.Done() first from readFull() function.
+func (t *rangeReaderTest) Test_ReadAt_ContextCancelAfterReadDoneSkipsCancel() {
+	t.rangeReader.reader = getReader(4)
+	t.rangeReader.start = 0
+	t.rangeReader.limit = 4
+	cancelCalled := false
+	t.rangeReader.cancel = func() {
+		cancelCalled = true
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	readReturned := make(chan struct{})
+
+	go func() {
+		_, _ = t.rangeReader.ReadAt(ctx, &gcsx.GCSReaderRequest{
+			Buffer:    make([]byte, 4),
+			Offset:    0,
+			EndOffset: 4,
+		})
+		close(readReturned)
+	}()
+	// Cancel the context to trigger ctx.Done()
+	go cancel()
+
+	// Let the read complete
+	select {
+	case <-readReturned:
+		// Pass
+	case <-time.After(100 * time.Millisecond):
+		t.T().Fatal("Expected read to finish before cancellation")
+	}
+	assert.False(t.T(), cancelCalled, "Expected rr.cancel NOT to be called since read already finished")
```

Comment:
This can fail, if cancel() method will be called before the read. I think, you need to make sure cancel() is called after the read-done.

What about just calling both ReadAt() and cancel() sequentially, instead of calling as background thread?

Reply:
So there are two cases and I wrote this test specifically for second code snippet
```
case <-readDone:
			return
```
```
**case <-ctx.Done():
			select {
			case <-readDone:
				return**

			default:
				rr.cancel()
			}
		}
```

If I do sequential call then it will go through the first code snippet on the other side if I trigger cancel first then it is not triggering any case. So to trigger this case I did parallelism.

On cancel call, I realized we are not doing `rr.cancel` in such case so cancel function will not being called, so I removed that assertion.

Reply:
As we discussed offline, making it serial also triggers the test with the race condition because the channel can listen in any sequence since its listener is triggered as a go routine. Therefore, I removed the parallelism.